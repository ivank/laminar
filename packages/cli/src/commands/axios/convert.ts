import { document, Document, mapWithContext, Type, Node, withImports, withIdentifier } from '@laminarjs/ts-compose';
import { oas31 } from 'openapi3-ts';
import ts from 'typescript';
import { convertSchema } from '../../convert-schema';
import { findMediaType } from '../../helpers';
import {
  AstContext,
  isSchemaObject,
  isResponseObject,
  getReferencedObject,
  isRequestBodyObject,
  isParameterObject,
} from '../../traverse';

const isTypeNode = (item: ts.TypeNode | undefined): item is ts.TypeNode => item !== undefined;
const isStatusOk = (status: string): boolean => Number(status) >= 200 && Number(status) <= 300;
const documentation = (summary?: string, description?: string): string | undefined =>
  summary || description ? [summary, description].filter((item) => item !== undefined).join('\n') : undefined;

const title = (str: string): string => str.replace(/^./, (first) => first.toUpperCase());
const cleanIdentifierName = (str: string): string => str.replace(/[^0-9a-zA-Z_$]+/g, '');
const pathToIdentifier = (method: string, path: string): string =>
  title(method) + path.split('/').map(cleanIdentifierName).map(title).join('');

const pathRegEx = /\{[^\}]+\}/g;

const pathToImplicitParams = (path: string): oas31.ParameterObject[] =>
  path.match(pathRegEx)?.map((name) => ({
    in: 'path',
    name: name.slice(1, -1),
    required: true,
    schema: { type: 'string' },
  })) ?? [];

const convertPathParams = (
  context: AstContext,
  path: string,
  parameters: oas31.ParameterObject[],
): { items: { name: string; type: ts.TypeNode }[]; context: AstContext } => {
  const implicitParams = pathToImplicitParams(path).map(
    (item) => parameters.find((param) => param.in === 'path' && param.name === item.name) ?? item,
  );

  return mapWithContext(context, implicitParams, (context, param) => {
    const astParams = param.schema ? convertSchema(context, param.schema) : document(context, Type.Unknown);

    return document(astParams?.context ?? context, {
      name: param.name,
      type: param.required ? astParams.type : Type.Union([astParams.type, Type.Undefined]),
    });
  });
};

const convertConfigProps = (
  context: AstContext,
  parameters: oas31.ParameterObject[],
): { items: ts.PropertySignature[]; context: AstContext } => {
  return mapWithContext(context, parameters, (context, param) => {
    const astParams = param.schema
      ? convertSchema({ ...context, optionalDefaults: true }, param.schema)
      : document(context, Type.Unknown);

    return document(
      { ...astParams.context, optionalDefaults: context.optionalDefaults },
      Type.Prop({
        name: param.in === 'header' ? param.name.toLowerCase() : param.name,
        type: astParams.type,
        jsDoc: documentation(param.description),
        isOptional: !param.required,
      }),
    );
  });
};

const convertConfigParams = (
  context: AstContext,
  parameters: oas31.ParameterObject[],
): Document<ts.PropertySignature[], AstContext> => {
  const headerItems = convertConfigProps(
    context,
    parameters.filter((param) => param.in === 'header'),
  );
  const queryItems = convertConfigProps(
    headerItems.context,
    parameters.filter((param) => param.in === 'query'),
  );

  return document(queryItems.context, [
    ...(headerItems.items.length
      ? [
          Type.Prop({
            name: 'headers',
            type: Type.TypeLiteral({ props: headerItems.items }),
            isOptional: true,
          }),
        ]
      : []),
    ...(queryItems.items.length
      ? [
          Type.Prop({
            name: 'params',
            type: Type.TypeLiteral({ props: queryItems.items }),
            isOptional: true,
          }),
        ]
      : []),
  ]);
};

const convertResponses = (
  context: AstContext,
  responses: oas31.ResponsesObject,
): Document<ts.TypeNode | undefined, AstContext> => {
  const responseEntries = Object.entries<oas31.ResponseObject | oas31.ReferenceObject>(responses)
    .filter(([status]) => status === 'default' || isStatusOk(status))
    .map(([, response]) => response);

  const astResponses = mapWithContext(context, responseEntries, (responseContext, responseOrRef) => {
    const response = getReferencedObject(responseOrRef, isResponseObject, 'response', responseContext);
    const schema = findMediaType('application/json', response?.content)?.schema;

    return schema ? convertSchema(responseContext, schema) : document(responseContext, undefined);
  });

  const responseTypes = astResponses.items.filter(isTypeNode);

  return document(
    astResponses.context,
    responseTypes.length === 0 ? undefined : responseTypes.length == 1 ? responseTypes[0] : Type.Union(responseTypes),
  );
};

const convertRequestBody = (
  context: AstContext,
  requestBodyOrRef?: oas31.RequestBodyObject | oas31.ReferenceObject,
): Document<{ isOptional: boolean; body: ts.TypeNode }, AstContext> => {
  const requestBody = requestBodyOrRef
    ? getReferencedObject(requestBodyOrRef, isRequestBodyObject, 'request-body', context)
    : undefined;
  const schema = findMediaType('application/json', requestBody?.content)?.schema;
  const convertedSchema = schema ? convertSchema(context, schema) : document(context, Type.Unknown);
  return document(convertedSchema.context, {
    isOptional: !requestBody?.required,
    body: convertedSchema.type,
  });
};

export const convertOapi = (context: AstContext, api: oas31.OpenAPIObject): Document<ts.Node, AstContext> => {
  const paths = mapWithContext(
    context,
    Object.entries<oas31.PathItemObject>(api?.paths ?? []),
    (pathContext, [path, pathApiOrRef]) => {
      const { parameters, description, summary, ...methodsApiOrRef } = pathApiOrRef;
      const pathApi = getReferencedObject(methodsApiOrRef, isSchemaObject, 'schema', context);

      const methods = mapWithContext(pathContext, Object.entries(pathApi), (methodContext, [method, operation]) => {
        const combinedParameters = [...(parameters ?? []), ...(operation.parameters ?? [])].map((item) =>
          getReferencedObject<oas31.ParameterObject>(item, isParameterObject, 'parameter', methodContext),
        );

        const requestName = `${method.toUpperCase()} ${path}`;
        const pathItems = convertPathParams(methodContext, path, combinedParameters);
        const configParams = convertConfigParams(pathItems.context, combinedParameters);

        const astRequestBody = convertRequestBody(configParams.context, operation.requestBody);

        const doc = documentation(operation.summary || summary, operation.description || description);

        const responseAst = convertResponses(astRequestBody.context, operation.responses);

        const hasData = method === 'post' || method === 'put' || method === 'patch';
        const configDataIdentifier = pathToIdentifier(method, path);
        const configDataContext = configParams.type.length
          ? withIdentifier(
              responseAst.context,
              Type.Interface({
                name: configDataIdentifier,
                props: configParams.type,
                isExport: true,
              }),
            )
          : responseAst.context;

        const node = Node.ObjectLiteralProp({
          key: requestName,
          jsDoc: doc,
          value: Node.Arrow({
            args: [
              ...pathItems.items.map(({ name }) => Type.Param({ name })),
              ...(hasData ? [Type.Param({ name: 'data' })] : []),
              Type.Param({ name: 'config' }),
            ],
            body: Node.Call({
              expression: Node.Identifier(`api.${method.toLowerCase()}`),
              typeArgs: responseAst.type ? [responseAst.type] : undefined,
              args: [
                Node.TemplateString(path.replace(/\{/g, '${')),
                ...(hasData ? [Node.Identifier('data')] : []),
                Node.Identifier('config'),
              ],
            }),
          }),
        });

        const type = Type.Prop({
          name: requestName,
          jsDoc: doc,
          type: Type.Arrow({
            args: [
              ...pathItems.items.map(Type.Param),
              ...(hasData
                ? [
                    Type.Param({
                      name: 'data',
                      type: astRequestBody.type.body,
                      isOptional: astRequestBody.type.isOptional,
                    }),
                  ]
                : []),
              Type.Param({
                name: 'config',
                isOptional: true,
                type: Type.Intersection([
                  Type.Referance('AxiosRequestConfig'),
                  ...(configParams.type.length ? [Type.Referance(configDataIdentifier)] : []),
                ]),
              }),
            ],
            ret: Type.Referance('Promise', [
              Type.Referance('AxiosResponse', responseAst.type ? [responseAst.type] : undefined),
            ]),
          }),
        });

        return document(configDataContext, { node, type });
      });

      return document(methods.context, methods.items);
    },
  );

  const jsDoc = [
    api.info.title,
    api.info.version ? `\n\nVersion: ${api.info.version}` : '',
    api.info.description ? `\n\nDescription:\n${api.info.description}` : '',
  ].join('');

  const endpoints = Node.ObjectLiteral({
    multiline: true,
    props: paths.items
      .reduce<ts.PropertyAssignment[]>((all, path) => all.concat(path.map((item) => item.node)), [])
      .concat(Node.ObjectLiteralProp({ key: 'api', value: Node.Identifier('api') })),
  });

  return document(
    withIdentifier(
      withImports(paths.context, {
        module: 'axios',
        named: [{ name: 'AxiosRequestConfig' }, { name: 'AxiosInstance' }, { name: 'AxiosResponse' }],
      }),
      Type.Interface({
        name: 'AxiosOapiInstance',
        isExport: true,
        props: paths.items
          .reduce<ts.PropertySignature[]>((all, path) => all.concat(path.map((item) => item.type)), [])
          .concat([Type.Prop({ name: 'api', type: Type.Referance('AxiosInstance') })]),
      }),
    ),
    Node.Const({
      name: 'axiosOapi',
      isExport: true,
      jsDoc,
      value: Node.Arrow({
        args: [Type.Param({ name: 'api', type: Type.Referance('AxiosInstance') })],
        ret: Type.Referance('AxiosOapiInstance'),
        body: endpoints,
      }),
    }),
  );
};
