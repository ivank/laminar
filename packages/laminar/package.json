{
  "name": "@ovotech/laminar",
  "version": "0.13.22",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "description": "A minimal nodejs http server, built around the concept of strictly typed middlewares",
  "author": "Ivan Kerin <ikerin@gmail.com>",
  "repository": "git@github.com:ovotech/laminar.git",
  "homepage": "https://github.com/ovotech/laminar/tree/main/packages/laminar#readme",
  "license": "Apache-2.0",
  "devDependencies": {
    "@ikerin/build-readme": "^1.1.1",
    "@types/cookie": "^0.5.4",
    "@types/jest": "^29.5.8",
    "@types/jsonwebtoken": "^9.0.5",
    "@types/mime-types": "^2.1.4",
    "@types/node": "^14.18.34",
    "@types/tough-cookie": "^4.0.5",
    "@types/uuid": "^9.0.7",
    "@typescript-eslint/eslint-plugin": "^5.10.1",
    "@typescript-eslint/parser": "^5.10.1",
    "axios": "^1.6.2",
    "axios-cookiejar-support": "^4.0.7",
    "csv-parse": "^5.5.2",
    "csv-stringify": "^6.4.4",
    "eslint": "^8.53.0",
    "eslint-config-prettier": "^8.3.0",
    "form-data": "^4.0.0",
    "jest": "^29.7.0",
    "jsonwebtoken": "^9.0.2",
    "nock": "^13.3.8",
    "prettier": "^2.5.1",
    "stream-mock": "^2.0.5",
    "tough-cookie": "^4.1.3",
    "ts-jest": "^29.1.1",
    "ts-node": "^10.9.1",
    "ts-retry-promise": "^0.6.0",
    "typescript": "^4.8.3",
    "uuid": "^9.0.1"
  },
  "scripts": {
    "build": "tsc --declaration",
    "build:readme": "build-readme README.md https://github.com/ovotech/laminar/tree/main/packages/laminar",
    "test": "jest test --runInBand",
    "lint": "F='{src,test}/**/*.ts' && prettier --check $F && eslint $F"
  },
  "jest": {
    "preset": "../../jest.json"
  },
  "dependencies": {
    "@ovotech/json-schema": "workspace:*",
    "mime-types": "^2.1.35",
    "openapi-schemas": "^2.1.0",
    "openapi3-ts": "^2.0.1"
  }
}
