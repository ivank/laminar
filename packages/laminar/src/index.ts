/**
 * @packageDocumentation
 * @module @laminarjs/laminar
 */
export {
  toIncomingMessageResolver,
  toRequestListener,
  HttpService,
  HttpServiceParams,
  HttpsServiceParams,
  IncomingMessageResolverParams,
  HttpParams,
} from './http/http-service';
export {
  optional,
  redirect,
  file,
  response,
  ok,
  noContent,
  movedPermanently,
  found,
  seeOther,
  notModified,
  badRequest,
  unauthorized,
  forbidden,
  notFound,
  internalServerError,
  json,
  yaml,
  form,
  binary,
  pdf,
  xml,
  text,
  html,
  css,
  csv,
  jsonOk,
  jsonCreated,
  jsonNoContent,
  jsonMovedPermanently,
  jsonFound,
  jsonSeeOther,
  jsonBadRequest,
  jsonUnauthorized,
  jsonForbidden,
  jsonNotFound,
  jsonInternalServerError,
  textOk,
  textCreated,
  textMovedPermanently,
  textFound,
  textSeeOther,
  textBadRequest,
  textUnauthorized,
  textForbidden,
  textNotFound,
  textInternalServerError,
  htmlOk,
  htmlCreated,
  htmlMovedPermanently,
  htmlFound,
  htmlSeeOther,
  htmlBadRequest,
  htmlUnauthorized,
  htmlForbidden,
  htmlNotFound,
  htmlInternalServerError,
  yamlOk,
  yamlCreated,
  yamlMovedPermanently,
  yamlFound,
  yamlSeeOther,
  yamlBadRequest,
  yamlUnauthorized,
  yamlForbidden,
  yamlNotFound,
  yamlInternalServerError,
  setCookie,
  SetCookie,
  ResponseCreator,
  FileOptions,
} from './http/response';
export { corsMiddleware, CorsConfig, OriginChecker } from './http/middleware/cors.middleware';
export {
  BodyParser,
  parseJson,
  parseForm,
  parseText,
  parseDefault,
  defaultBodyParsers,
  bodyParserMiddleware,
  concatStream,
  parseBody,
} from './http/middleware/body-parser.middleware';
export { toHttpRequest } from './http/request';
export { parseCookie, serializeCookie, CookieOptions } from './http/cookie';
export { parseQueryObjects, toJson, Json } from './helpers';
export {
  responseTimeMiddleware,
  ResponseTimeConfig,
  defaultResponseTimeHeader,
} from './http/middleware/response-time.middleware';
export { requestLoggingMiddleware, RequestLoggingMiddlewareParams } from './http/middleware/request-logging.middleware';

export { LoggerMetadata, LoggerLike, withStaticMetadata, LoggerContext, loggerMiddleware } from './logger/index';

export { start, stop, Application, init, run, InitOrder, StartLogger, StopLogger } from './application';
export {
  responseParserMiddleware,
  defaultResponseParsers,
  jsonResponseParser,
  formResponseParser,
  ResponseParser,
  parseResponse,
} from './http/middleware/response-parser.middleware';
export {
  errorsMiddleware,
  defaultErrorHandler,
  RequestError,
  HttpErrorHandler,
} from './http/middleware/errors.middleware';
export {
  toMultipartBoundary,
  toMultipartData,
  parseMultipartBuffer,
  MultipartData,
  MultipartItem,
  MultipartParser,
} from './http/multipart-parser';
export {
  HttpContext,
  HttpResponseBody,
  HttpResponse,
  HttpListener,
  HttpMiddleware,
  IncomingMessageResolver,
} from './http/types';
export { HttpError, isHttpError } from './http/http-error';
export {
  router,
  get,
  post,
  patch,
  del,
  options,
  put,
  route,
  staticAssets,
  RouteContext,
  AppRoute,
  PathRoute,
  PathRouteOptions,
  StaticAssetsOptions,
  Matcher,
} from './http/router';
export { Empty, Service, AbstractMiddleware, Middleware } from './types';
export {
  openApi,
  securityOk,
  securityError,
  securityRedirect,
  isSecurityOk,
  defaultError,
  OapiContext,
  AppRouteOapi,
  OapiPath,
  OapiAuthInfo,
  SecurityOk,
  Security,
  RequestSecurityResolver,
  OapiSecurityResolver,
  OapiSecurity,
  ResponseOapi,
  OapiPaths,
  OapiConfig,
} from './http/open-api';

export { toErrorMetadata, LaminarError } from './errors';
export { passThroughMiddleware } from './middleware/pass-through.middleware';
