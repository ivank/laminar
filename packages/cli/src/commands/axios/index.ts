import { Command, createCommand, Option } from 'commander';
import fs from 'fs';
import { printDocument } from '@laminar/ts-compose';
import { convertOapi } from './convert';
import chalk from 'chalk';
import { dirname } from 'path';
import { Logger } from '../../types';
import { toContext, concatStreamToString, parseSchema } from '../../helpers';

export const processFile = async (file: string): Promise<string> => {
  const { context, value } = await toContext(file);
  return printDocument(convertOapi(context, value));
};

export const axiosCommand = (logger: Logger = console): Command =>
  createCommand('axios')
    .description('Convert openapi schemas to typescript types for axios')
    .option('-f, --file <file>', 'Filename to process, uses STDIN if not specified')
    .addOption(new Option('-t, --stdin-type <type>', 'Type of STDIN data').default('json').choices(['json', 'yaml']))
    .option('-o, --output <output>', 'File to output to, uses STDOUT if not specified')
    .action(async ({ file, output, stdinType }) => {
      const input = file ? file : parseSchema(stdinType, await concatStreamToString(process.stdin));
      const result = await processFile(input);

      if (output) {
        fs.mkdirSync(dirname(output), { recursive: true });
        logger.info(`OpanAPI Schema ${chalk.green(file ? file : 'STDIN')} -> ${chalk.yellow(output)} axios types`);
        fs.writeFileSync(output, result);
      } else {
        process.stdout.write(result);
      }
    });
