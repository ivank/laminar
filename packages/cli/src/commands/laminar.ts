import commander from 'commander';
import { apiCommand } from './api';
import { axiosCommand } from './axios';
import { jsonSchemaCommand } from './json-shema';
import { convertCommand } from './avro';
import { Logger } from '../types';
import { fixturesCommand } from './fixtures';

export const laminarCommand = (output: Logger = console): commander.Command =>
  commander
    .createCommand('laminar')
    .version('0.15.0')
    .description(
      `Laminar CLI - a tool for converting OpenApi and Avro Schemas to appropriate typescript types

laminar api
  Convert openapi schema to types used by laminar api (openApiTyped middleware)

laminar axios
  Convert openapi schema to types used by axios

laminar json-schema
  Convert json schema to typescript types

laminar avro
  Convert avro to typescript types

laminar fixtures
  Load pg tables and generate @laminar/fixtures types from them

Example:
  laminar api --file openapi.yaml --output src/__generated__/openapi.yaml.ts
  curl http://example.com/openapi.yaml | laminar api --stdin-type yaml | prettier --stdin-filepath openapi.yamlts > src/__generated__/openapi.yaml.ts
  laminar api --file openapi.yaml --output src/__generated__/openapi.yaml.ts --watch
  laminar avro path/to/*.avsc --output-dir src/__generated__ --logical-type timestamp-millis=Date
  laminar fixtures postgres://example-admin:example-pass@localhost:5432/example --output src/__generated__/fixtures.ts
`,
    )
    .addCommand(apiCommand(output))
    .addCommand(axiosCommand(output))
    .addCommand(jsonSchemaCommand(output))
    .addCommand(convertCommand(output))
    .addCommand(fixturesCommand(output));
