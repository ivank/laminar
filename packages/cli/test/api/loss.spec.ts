import { HttpService, jsonOk, run } from '@laminar/laminar';
import axios from 'axios';
import { join } from 'path';
import { openApiTyped } from './__generated__/loss';

describe('Statements', () => {
  it('Should process response', async () => {
    const listener = await openApiTyped({
      api: join(__dirname, 'loss.yaml'),
      paths: {
        '/accounts/{accountId}/meters': {
          get: async () =>
            jsonOk([
              {
                meterType: 'gas',
                mpxn: '111111111',
                plan: 'PA_G_WM_R_25',
                tariffName: 'SMART PAYG (all online)',
                startDate: '2019-01-04',
                endDate: '2019-06-02',
                msn: 'G4111111',
                ldz: 'WM',
                address: '87, TEST AVENUE, TEST TOWN',
                postCode: 'TT123EE',
              },
              {
                meterType: 'elec',
                mpxn: '14111111',
                plan: 'PA_E_14_R_1_25',
                tariffName: 'SMART PAYG (all online)',
                startDate: '2019-01-04',
                endDate: '2019-06-02',
                msn: '222222222',
                profileClassId: '01',
                meterTimeswitchCode: '801',
                lineLossFactorClassId: '1',
                address: '87, TEST AVENUE, LOW HILL, TEST TOWN, WEST MIDLANDS',
                postCode: 'TT123EE',
              },
            ]),
        },
      },
    });
    const http = new HttpService({ listener, port: 4910 });
    await run({ initOrder: [http] }, async () => {
      const api = axios.create({ baseURL: 'http://localhost:4910' });
      const { data } = await api.get('/accounts/123/meters');

      expect(data).toMatchSnapshot();
    });
  });
});
