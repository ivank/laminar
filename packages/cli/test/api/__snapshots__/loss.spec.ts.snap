// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Statements Should process response 1`] = `
[
  {
    "address": "87, TEST AVENUE, TEST TOWN",
    "endDate": "2019-06-02",
    "ldz": "WM",
    "meterType": "gas",
    "mpxn": "111111111",
    "msn": "G4111111",
    "plan": "PA_G_WM_R_25",
    "postCode": "TT123EE",
    "startDate": "2019-01-04",
    "tariffName": "SMART PAYG (all online)",
  },
  {
    "address": "87, TEST AVENUE, LOW HILL, TEST TOWN, WEST MIDLANDS",
    "endDate": "2019-06-02",
    "lineLossFactorClassId": "1",
    "meterTimeswitchCode": "801",
    "meterType": "elec",
    "mpxn": "14111111",
    "msn": "222222222",
    "plan": "PA_E_14_R_1_25",
    "postCode": "TT123EE",
    "profileClassId": "01",
    "startDate": "2019-01-04",
    "tariffName": "SMART PAYG (all online)",
  },
]
`;
